src/ring_theory/polynomial/symmetric.lean

```diff
@@ -0,0 +1,155 @@
+
+This file defines symmetric polynomials and elementary symmetric polynomials.
+We also prove some basic facts about them.
+
+## Main declarations
+
+* `mv_polynomial.is_symmetric`
+
+* `mv_polynomial.esymm`
+
+## Notation
+
++ `esymm σ R n`, is the `n`th elementary symmetric polynomial in `mv_polynomial σ R`.
+
+As in other polynomial files, we typically use the notation:
+
++ `σ τ : Type*` (indexing the variables)
+
++ `R S : Type*` `[comm_semiring R]` `[comm_semiring S]` (the coefficients)
+
++ `r : R` elements of the coefficient ring
+
++ `i : σ`, with corresponding monomial `X i`, often denoted `X_i` by mathematicians
+
++ `φ ψ : mv_polynomial σ R`
+
+-/
+
+open equiv (perm)
+open_locale big_operators
+noncomputable theory
+
+namespace mv_polynomial
+
+variables {σ : Type*} {R : Type*}
+variables {τ : Type*} {S : Type*}
+
+/-- A mv_polynomial φ is symmetric if it is invariant under
+permutations of its variables by the  `rename` operation -/
+def is_symmetric [comm_semiring R] (φ : mv_polynomial σ R) : Prop :=
+∀ e : perm σ, rename e φ = φ
+
+namespace is_symmetric
+variables [comm_semiring R] [comm_semiring S] {φ ψ : mv_polynomial σ R}
+
+@[simp]
+lemma C (r : R) : is_symmetric (C r : mv_polynomial σ R) :=
+λ e, rename_C e r
+
+@[simp]
+lemma zero : is_symmetric (0 : mv_polynomial σ R) :=
+by { rw [← C_0], exact is_symmetric.C 0 }
+
+@[simp]
+lemma one : is_symmetric (1 : mv_polynomial σ R) :=
+by { rw [← C_1], exact is_symmetric.C 1 }
+
+lemma add (hφ : is_symmetric φ) (hψ : is_symmetric ψ) : is_symmetric (φ + ψ) :=
+λ e, by rw [alg_hom.map_add, hφ, hψ]
+
+lemma mul (hφ : is_symmetric φ) (hψ : is_symmetric ψ) : is_symmetric (φ * ψ) :=
+λ e, by rw [alg_hom.map_mul, hφ, hψ]
+
+lemma smul (r : R) (hφ : is_symmetric φ) : is_symmetric (r • φ) :=
+λ e, by rw [alg_hom.map_smul, hφ]
+
+@[simp]
+lemma map (hφ : is_symmetric φ) (f : R →+* S) : is_symmetric (map f φ) :=
+λ e, by rw [← map_rename, hφ]
+
+end is_symmetric
+
+namespace is_symmetric
+
+variables [comm_ring R] {φ ψ : mv_polynomial σ R}
+
+lemma neg (hφ : is_symmetric φ) : is_symmetric (-φ) :=
+λ e, by rw [alg_hom.map_neg, hφ]
+
+lemma sub (hφ : is_symmetric φ) (hψ : is_symmetric ψ) : is_symmetric (φ - ψ) :=
+λ e, by rw [alg_hom.map_sub, hφ, hψ]
+
+end is_symmetric
+
+section elementary_symmetric
+
+open finset
+
+variables (σ R) [comm_semiring R] [comm_semiring S] [fintype σ] [fintype τ]
+
+/-- Define `esymm σ R n` as the `n`th elementary symmetric polynomial in `mv_polynomial σ R`. -/
+def esymm (n : ℕ) : mv_polynomial σ R :=
+∑ t in powerset_len n univ, ∏ i in t, X i
+
+/-- A second definition of `esymm σ R n` that ranges over a subtype instead of `powerset_len`. -/
+lemma esymm₂ (n : ℕ) : (esymm σ R n) =
+  ∑ t : {s : finset σ // s.card = n}, ∏ i in (t : finset σ), X i :=
+begin
+  rw esymm,
+  let i : Π (a : finset σ), a ∈ powerset_len n univ → {s : finset σ // s.card = n} :=
+  by {intros a ha, split, exact (mem_powerset_len.mp ha).2, apply_instance },
```

Comment:
We don't need tactics to define this data, term mode gives slightly more readable (to me) results:
```suggestion
  let i : Π (a : finset σ), a ∈ powerset_len n univ → {s : finset σ // s.card = n} :=
    λ a ha, ⟨_, (mem_powerset_len.mp ha).2⟩,
```